lib/puppet/ssl/state_machine.rb

```diff
@@ -349,6 +362,36 @@ def next_state
     end
   end
 
+  # Class to renew a client/host certificate automatically.
+  #
+  class NeedRenewedCert < KeySSLState
+    def next_state
+      Puppet.debug(_("Renewing client certificate"))
+
+      route = @machine.session.route_to(:ca, ssl_context: @ssl_context)
+      cert = OpenSSL::X509::Certificate.new(
+        route.post_certificate_renewal(@ssl_context)[1]
+      )
+
+      # verify client cert before saving
+      next_ctx = @ssl_provider.create_context(
+        cacerts: @ssl_context.cacerts, crls: @ssl_context.crls, private_key: @private_key, client_cert: cert
+      )
+      @cert_provider.save_client_cert(Puppet[:certname], cert)
+
+      Puppet.info("Refreshed client certificate: #{@machine.digest_as_hex(cert.to_pem)}, not before '#{cert.not_before}', not after '#{cert.not_after}'")
+      
+      Done.new(@machine, next_ctx)
+    rescue Puppet::HTTP::ResponseError => e
+      if e.response.code == 404
+        Puppet.info(_("Certificate autorenewal has not been enabled on the server."))
+      else
+        Puppet.warning(_("Failed to automatically renew certificate: %{code} %{reason}") % { code: e.response.code, reason: e.response.reason })
+      end
+      Done.new(@machine, @ssl_context)
+    end
```

Comment:
Ah one thing I missed is we may receive other exceptions, e.g. can't connect to the CA server, etc, so I think we'll want a "catch all" rescue after the `ResponseError`, and the log a warning and return `Done` like:

```ruby
rescue Puppet::HTTP::ResponseError => e
  ...
rescue => e
  ...
end
```

For example, if you force the client to renew but ask it to connect to a bogus CA:

```
$ bundle exec puppet agent -t --ca_server ca.example.com
Error: Connection to https://ca.example.com:8140/puppet-ca/v1 failed, trying next route: Request to https://ca.example.com:8140/puppet-ca/v1 failed after 0.238 seconds: Failed to open TCP connection to ca.example.com:8140 (getaddrinfo: Name or service not known)
Wrapped exception:
Failed to open TCP connection to ca.example.com:8140 (getaddrinfo: Name or service not known)
Error: No more routes to ca
```

The other states don't do that, because those kinds of errors cause us to go to the `Wait` state and try again later.